src/dbnode/network/server/tchannelthrift/node/service.go

```diff
@@ -256,22 +263,49 @@ func (s *service) Bootstrapped(ctx thrift.Context) (*rpc.NodeBootstrappedResult_
 }
 
 func (s *service) Query(tctx thrift.Context, req *rpc.QueryRequest) (*rpc.QueryResult_, error) {
+	ctx := tchannelthrift.Context(tctx)
+
+	var (
+		sp    opentracing.Span
+		spCtx stdctx.Context
+	)
+
+	goCtx, exists := ctx.GoContext()
+	if exists {
+		sp, spCtx = opentracingutil.StartSpanFromContext(goCtx, querySpanID)
+		sp.LogFields(
+			opentracinglog.String("query", req.Query.String()),
+			opentracinglog.String("namespace", req.NameSpace),
+			opentracingutil.Time("start", time.Unix(0, req.RangeStart)),
+			opentracingutil.Time("end", time.Unix(0, req.RangeStart)),
+		)
+
+		ctx.SetGoContext(spCtx)
+		defer sp.Finish()
+	}
+
 	if s.isOverloaded() {
 		s.metrics.overloadRejected.Inc(1)
 		return nil, tterrors.NewInternalError(errServerIsOverloaded)
 	}
 
-	ctx := tchannelthrift.Context(tctx)
-
 	start, rangeStartErr := convert.ToTime(req.RangeStart, req.RangeType)
 	end, rangeEndErr := convert.ToTime(req.RangeEnd, req.RangeType)
 
 	if rangeStartErr != nil || rangeEndErr != nil {
-		return nil, tterrors.NewBadRequestError(xerrors.FirstError(rangeStartErr, rangeEndErr))
+		err := xerrors.FirstError(rangeStartErr, rangeEndErr)
+		if sp != nil {
+			sp.LogFields(opentracinglog.Error(err))
+		}
+
+		return nil, tterrors.NewBadRequestError(err)
 	}
 
 	q, err := convert.FromRPCQuery(req.Query)
 	if err != nil {
+		if sp != nil {
+			sp.LogFields(opentracinglog.Error(err))
```

Comment:
I think you can prevent having to do this everywhere by making the return arguments of this function named, so it becomes:

```golang
func (s *service) Query(tctx thrift.Context, req *rpc.QueryRequest) (result *rpc.QueryResult_, err error)
```

and then after you create the span you can do something like

```golang
defer func() {
    if err != nil && sp != nil {
        sp.LogFields(opentracinglog.Error(err))
   }
}()
```

This shouldn't interfere with the call to `defer sp.Finish()` since defers are executed last-in first-out order

Reply:
The easier thing, is just wrap this in another method.

i.e. making `Query` wrap calls to an inner `query`

```go
func (s *service) Query(...) (...) {
  var span opentracing.Span

  // maybe create span based on whether this is being sampled or not

  // call the actual method, with reference to the span)
  result, err := s.query(..., span)

  // now optionally add the error
  if span != nil && err != nil {
    span.LogFields(opentracinglog.Error(err)
  }

  // Also don't need to use a defer now to close the span
  span.Finish()

  return result, err
}